core/src/main/java/org/kiwix/kiwixmobile/core/history/HistoryAdapter.kt

```diff
@@ -0,0 +1,149 @@
+import org.threeten.bp.LocalDate
+import org.threeten.bp.format.DateTimeFormatter
+
+internal class HistoryAdapter(
+  val historyList: List<HistoryListItem>,
+  val deleteList: List<HistoryListItem>,
+  val itemClickListener: OnItemClickListener
+) : Adapter<ViewHolder>() {
+
+  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+    return if (viewType == TYPE_ITEM) {
+      Item(inflateLayoutFromParent(parent, layout.item_bookmark_history))
+    } else {
+      Category(inflateLayoutFromParent(parent, layout.header_date))
+    }
+  }
+
+  private fun inflateLayoutFromParent(parent: ViewGroup, layoutId: Int) =
+    LayoutInflater.from(parent.context)
+      .inflate(layoutId, parent, false)
+
+  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+    if (holder is Item) {
+      val history = historyList[position] as HistoryItem
+      setItemDataWithHelpOfHistoryItem(holder, history)
+    } else if (holder is Category) {
+      val date = (historyList[position] as DateItem).dateString
+      setCategoryDataWithHelpOfDate(date, holder)
+    }
+  }
+
+  private fun setCategoryDataWithHelpOfDate(date: String, holder: Category) {
+    val formatter = DateTimeFormatter.ofPattern("d MMM yyyy")
+    val todaysDate = LocalDate.now().format(formatter)
+    val yesterdayDate = LocalDate.now().minusDays(1).format(formatter)
+
+    if (todaysDate != null && todaysDate.contentEquals(date)) {
+      holder.date.setText(
+        R.string.time_today
+      )
+    } else if (yesterdayDate != null && yesterdayDate.contentEquals(date)) {
+      holder.date.setText(
+        R.string.time_yesterday
+      )
+    } else {
+      holder.date.text = date
+    }
+  }
+
+  private fun setItemDataWithHelpOfHistoryItem(holder: Item, history: HistoryItem) {
+    holder.title.text = history.historyTitle
+    if (deleteList.contains(history)) {
+      holder.favicon.setImageDrawable(
+        ContextCompat.getDrawable(
+          holder.favicon.context,
+          drawable.ic_check_circle_blue_24dp
+        )
+      )
+    } else {
+      holder.favicon.setBitmapFromString(history.favicon)
+    }
+    holder.itemView.setOnClickListener {
+      itemClickListener.onItemClick(
+        holder.favicon, history
+      )
+    }
+    holder.itemView.setOnLongClickListener {
+      itemClickListener.onItemLongClick(
+        holder.favicon, history
+      )
+    }
+  }
+
+  override fun getItemViewType(position: Int): Int {
+    return if (historyList[position] is DateItem) 0 else TYPE_ITEM
+  }
+
+  override fun getItemCount(): Int {
+    return historyList.size
+  }
+
+  internal interface OnItemClickListener {
+    fun onItemClick(
+      favicon: ImageView,
+      history: HistoryItem
+    )
+
+    fun onItemLongClick(
+      favicon: ImageView,
+      history: HistoryItem
+    ): Boolean
+  }
+
+  internal inner class Item(itemView: View) : ViewHolder(itemView) {
+    @BindView(id.favicon) lateinit var favicon: ImageView
```

Comment:
Don't use Butterknife in kotlin, use viewbinding from kotlin android extensions.

create bind methods for these viewholders and move them to a separate class

Reply:
I tried using view bindings but the tutorial I followed used an inflated automatically created binding class for an activity which I can not seem to do for a custom view. Can you point me in the right direction for how to create a binding class for item bookmark history?

Reply:
Look at any ViewHolder in kotlin in the project.

That is the wrong ViewBindings, confusing I know.

We use [kotlin android extensions view binding](https://kotlinlang.org/docs/tutorials/android-plugin.html#view-binding), what you may have been looking at was the [new ViewBinding from Google](https://developer.android.com/topic/libraries/view-binding)

Reply:
Ah, thanks!

Reply:
like this?

```
internal inner class Item(itemView: View) : BaseViewHolder<View>(itemView) {
    val favicon: ImageView = itemView.favicon
    val title: TextView = itemView.title

    override fun bind(item: View) {
    }
}
```

Reply:
You don't need to assign to fields or access the views by the itemview.

in `bind` just use the ids from the xml

```
  class DownloadViewHolder(view: View, private val clickAction: (LibraryDownloadItem) -> Unit) :
    LibraryViewHolder<LibraryDownloadItem>(view) {

    override fun bind(item: LibraryDownloadItem) {
      libraryDownloadFavicon.setBitmap(item.favIcon)
      libraryDownloadTitle.text = item.title
      libraryDownloadDescription.text = item.description
      downloadProgress.progress = item.progress
      stop.setOnClickListener { clickAction.invoke(item) }
      downloadState.text = item.downloadState.toReadableState(containerView.context)
      eta.text = item.readableEta
    }
  }
```

Reply:
Also `BaseViewHolder` takes a type parameter. Depending on the Type Parameter you specify it means you will have to implement `bind(item:TypeParamter)`

Reply:
Think I got it  now! I probably got confused because the TextView with id = `header_date` had a variable name of `date` which made it a bit confusing when trying to set the text attribute or trying to access date from the ViewHolder. Thanks for your feedback and help.